tket/src/Transformations/Decomposition.cpp

```diff
@@ -1815,6 +1815,178 @@ Transform decomp_arbitrary_controlled_gates() {
   return rebase_factory(allowed_gate_set, CircPool::CX(), CircPool::tk1_to_tk1);
 }
 
+static void substitute_cnx(
+    const Command &cmd, const std::optional<qubit_vector_t> &new_args,
+    const bool &as_dagger, Circuit &circ,
+    std::map<unsigned, Circuit> &cnx_cache) {
+  Circuit cnx;
+  qubit_vector_t original_args = cmd.get_qubits();
+  bool conditional = cmd.get_op_ptr()->get_type() == OpType::Conditional;
+
+  // Compute the decomposition
+  auto cache = cnx_cache.find((unsigned)original_args.size());
+  if (cache != cnx_cache.end()) {
+    cnx = cache->second;
+  } else {
+    cnx = multi_controlled_to_2q(
+        get_op_ptr(OpType::CnX, std::vector<Expr>(), original_args.size()));
+    cnx_cache.insert({(unsigned)original_args.size(), cnx});
+  }
+
+  // Reorder the replacement circuit by renaming the qubits
+  if (new_args != std::nullopt) {
+    unit_map_t qmap;
+    for (unsigned i = 0; i + 1 < new_args.value().size(); i++) {
+      auto it =
+          find(original_args.begin(), original_args.end(), new_args.value()[i]);
+      int index = it - original_args.begin();
+      qmap.insert({Qubit(i), Qubit(index)});
+    }
+    cnx.rename_units(qmap);
+  }
+
+  if (as_dagger) {
+    cnx = cnx.dagger();
+  }
+  if (conditional) {
+    circ.substitute_conditional(
+        cnx, cmd.get_vertex(), Circuit::VertexDeletion::No);
+  } else {
+    circ.substitute(cnx, cmd.get_vertex(), Circuit::VertexDeletion::No);
+  }
+}
+
+Transform cnx_pairwise_decomposition() {
+  return Transform([](Circuit &circ) {
+    bool success = commute_through_multis().apply(circ);
+    success |= remove_redundancies().apply(circ);
+
+    // Cache CnX decompositions
+    std::map<unsigned, Circuit> cnx_cache;
+    // Replaced vertices to delete at the end
+    VertexList bin;
+
+    // Find all CnX gates, including conditional ones
+    std::vector<Command> commands = circ.get_commands();
```

Comment:
I wonder if in the future we can make runtime improvements by avoiding flattening the DAG and just iterating through graph vertices until a match is found. We can shelve it for now but it might be worth adding a very low priority TKET task for this.

Reply:
Yeah, I agree. I think we can also consider how to pick an _optimal_ decomposition if multiple matches are found (e.g. multiple CnX, CX, etc. neighbouring a CnX).

Reply:
Makes sense. You talked before about potentially producing a new IR for doing better decompositions of sequences of multi-controlled gates (plausibly some kind of "`MultiplexorGraph`"), so maybe we try to assign some time in the future for considering that.